pkg/agent/controller/packetsampling/packetsampling_controller.go

```diff
@@ -0,0 +1,759 @@
+	return os.Remove(uidToPath(uid))
+}
+
+func uidToPath(uid string) string {
+	return path.Join(packetDirectory, uid+".pcapng")
+}
+
+func (c *Controller) worker() {
+	for c.processPacketSamplingItem() {
+
+	}
+}
+
+func (c *Controller) processPacketSamplingItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncPacketSampling(key); err == nil {
+		c.queue.Forget(key)
+	} else {
+		klog.Errorf("Error syncing PacketSampling %s, existing. Error: %v", key, err)
+	}
+	return true
+}
+
+func (c *Controller) validatePacketSampling(ps *crdv1alpha1.PacketSampling) error {
+	if ps.Spec.Destination.Service != "" && !c.enableAntreaProxy {
+		return errors.New("using Service destination requires AntreaPolicy feature gate")
+	}
+
+	if ps.Spec.Destination.IP != "" {
+		destIP := net.ParseIP(ps.Spec.Destination.IP)
+		if destIP == nil {
+			return fmt.Errorf("destination IP %s is not valid", ps.Spec.Destination.IP)
+		}
+		if !c.enableAntreaProxy && c.serviceCIDR.Contains(destIP) {
+			return errors.New("using ClusterIP destination requires AntreaPolicy feature gate")
+		}
+
+	}
+	return nil
+}
+
+func (c *Controller) errorPacketSamplingCRD(ps *crdv1alpha1.PacketSampling, reason string) (*crdv1alpha1.PacketSampling, error) {
+	ps.Status.Phase = crdv1alpha1.PacketSamplingFailed
+
+	type PacketSampling struct {
+		Status crdv1alpha1.PacketSamplingStatus `json:"status,omitempty"`
+	}
+
+	patchData := PacketSampling{
+		Status: crdv1alpha1.PacketSamplingStatus{Phase: ps.Status.Phase, Reason: reason},
+	}
+	payloads, _ := json.Marshal(patchData)
+	return c.crdClient.CrdV1alpha1().PacketSamplings().Patch(context.TODO(), ps.Name, types.MergePatchType, payloads, metav1.PatchOptions{}, "status")
+
+}
+
+func (c *Controller) cleanupPacketSampling(psName string) {
+	psState := c.deletePacketSamplingState(psName)
+	if psState != nil {
+		err := c.ofClient.UninstallPacketSamplingFlows(psState.tag)
+		if err != nil {
+			klog.Errorf("Error cleaning up flows for PacketSampling %s: %v", psName, err)
+		}
+
+		err = psState.pcapngWriter.Flush()
+		if err != nil {
+			klog.Errorf("Error flushing pcapng file for PacketSampling %s: %v", psName, err)
+		}
+		err = psState.pcapngFile.Close()
+		if err != nil {
+			klog.Errorf("Error closing pcapng file for PacketSampling %s: %v", psName, err)
+		}
+	}
+}
+
+func (c *Controller) deletePacketSamplingState(psName string) *packetSamplingState {
+	c.runningPacketSamplingsMutex.Lock()
+	defer c.runningPacketSamplingsMutex.Unlock()
+
+	for tag, state := range c.runningPacketSamplings {
+		if state.name == psName {
+			delete(c.runningPacketSamplings, tag)
+			return state
+		}
+	}
+	return nil
+}
+
+func (c *Controller) startPacketSampling(ps *crdv1alpha1.PacketSampling) error {
+	err := c.validatePacketSampling(ps)
+
+	func() {
```

Comment:
Why do you need a 'func()' here? you mean `defer func()`

Reply:
updated